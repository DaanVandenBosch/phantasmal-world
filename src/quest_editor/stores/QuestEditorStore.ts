import { property, list_property } from "../../core/observable";
import { QuestModel } from "../model/QuestModel";
import { Property, PropertyChangeEvent } from "../../core/observable/property/Property";
import { read_file } from "../../core/read_file";
import { parse_quest, write_quest_qst } from "../../core/data_formats/parsing/quest";
import { ArrayBufferCursor } from "../../core/data_formats/cursor/ArrayBufferCursor";
import { Endianness } from "../../core/data_formats/Endianness";
import { QuestObjectModel } from "../model/QuestObjectModel";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { AreaModel } from "../model/AreaModel";
import { area_store } from "./AreaStore";
import { SectionModel } from "../model/SectionModel";
import { QuestEntityModel } from "../model/QuestEntityModel";
import { Disposable } from "../../core/observable/Disposable";
import { Disposer } from "../../core/observable/Disposer";
import { gui_store, GuiTool } from "../../core/stores/GuiStore";
import { UndoStack } from "../../core/undo/UndoStack";
import { TranslateEntityAction } from "../actions/TranslateEntityAction";
import { EditShortDescriptionAction } from "../actions/EditShortDescriptionAction";
import { EditLongDescriptionAction } from "../actions/EditLongDescriptionAction";
import { EditNameAction } from "../actions/EditNameAction";
import { EditIdAction } from "../actions/EditIdAction";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { create_new_quest } from "./quest_creation";
import { CreateEntityAction } from "../actions/CreateEntityAction";
import { RemoveEntityAction } from "../actions/RemoveEntityAction";
import { Euler, Vector3 } from "three";
import { RotateEntityAction } from "../actions/RotateEntityAction";
import { convert_quest_from_model, convert_quest_to_model } from "./model_conversion";
import Logger = require("js-logger");
import { MessageLogStore, LogMessage, LogLevel, LogGroup } from "../../core/gui/MessageLog";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { WritableProperty } from "../../core/observable/property/WritableProperty";
import { QuestRunner } from "../QuestRunner";

const logger = Logger.get("quest_editor/gui/QuestEditorStore");

export interface Logger {
    debug(...items: any[]): void;
    info(...items: any[]): void;
    warning(...items: any[]): void;
    error(...items: any[]): void;
}

export class QuestEditorStore implements Disposable, MessageLogStore {
    private readonly disposer = new Disposer();
    public readonly quest_runner: QuestRunner = new QuestRunner();
    /**
     * Log levels as Record<name, LogLevel>
     */
    private readonly log_levels_map = (function<KT extends readonly string[]>(names: KT) {
        type K = KT[keyof KT & number];
        return names.reduce<Record<K, LogLevel>>(
            (accum, name: K, idx) => {
                accum[name] = {
                    name: name,
                    value: idx,
                };
                return accum;
            },
            {} as Record<K, LogLevel>,
        );
    })(
        // Log level names in order of importance
        ["Debug", "Info", "Warning", "Error"] as const,
    );

    /**
     * Log levels as LogLevel[]
     */
    readonly log_levels = Object.values(this.log_levels_map);
    private readonly default_log_level = this.log_levels_map.Info;

    private log_group_key_counter = 0;
    private readonly _log_groups = list_property<LogGroup>();
    private readonly default_log_group: LogGroup = this.add_log_group("Default");

    private readonly _current_quest_filename = property<string | undefined>(undefined);
    private readonly _current_quest = property<QuestModel | undefined>(undefined);
    private readonly _current_area = property<AreaModel | undefined>(undefined);
    private readonly _selected_entity = property<QuestEntityModel | undefined>(undefined);
    private readonly _log_messages = list_property<LogMessage>();
    private readonly _log_level = property<LogLevel>(this.default_log_level);
    private readonly _log_group = property<LogGroup>(this.default_log_group);
    private readonly _show_all_log_groups = property<boolean>(true);

    readonly debug: WritableProperty<boolean> = property(false);
    readonly undo = new UndoStack();
    readonly current_quest_filename: Property<string | undefined> = this._current_quest_filename;
    readonly current_quest: Property<QuestModel | undefined> = this._current_quest;
    readonly current_area: Property<AreaModel | undefined> = this._current_area;
    readonly selected_entity: Property<QuestEntityModel | undefined> = this._selected_entity;
    readonly log_messages: ListProperty<LogMessage>;
    readonly log_level: Property<LogLevel> = this._log_level;
    readonly log_group: Property<LogGroup> = this._log_group;
    readonly log_groups: ListProperty<LogGroup> = this._log_groups;

    constructor() {
        this.disposer.add_all(
            gui_store.tool.observe(
                ({ value: tool }) => {
                    if (tool === GuiTool.QuestEditor) {
                        this.undo.make_current();
                    }
                },
                { call_now: true },
            ),

            this.current_quest
                .flat_map(quest => (quest ? quest.npcs : property([])))
                .observe(({ value: npcs }) => {
                    const selected = this.selected_entity.val;

                    if (selected instanceof QuestNpcModel && !npcs.includes(selected)) {
                        this.set_selected_entity(undefined);
                    }
                }),

            this.current_quest
                .flat_map(quest => (quest ? quest.objects : property([])))
                .observe(({ value: objects }) => {
                    const selected = this.selected_entity.val;

                    if (selected instanceof QuestObjectModel && !objects.includes(selected)) {
                        this.set_selected_entity(undefined);
                    }
                }),

            // force refresh
            this._log_level.observe(() => (this._log_messages.val = this._log_messages.val)),
            this._log_group.observe(() => (this._log_messages.val = this._log_messages.val)),
            this._show_all_log_groups.observe(
                () => (this._log_messages.val = this._log_messages.val),
            ),
        );

        this.log_messages = this._log_messages.filtered(this.log_message_predicate);
    }

    dispose(): void {
        this.disposer.dispose();
    }

    set_current_area = (area?: AreaModel): void => {
        this._selected_entity.val = undefined;

        this._current_area.val = area;
    };

    set_selected_entity = (entity?: QuestEntityModel): void => {
        if (entity && this.current_quest.val) {
            this._current_area.val = area_store.get_area(
                this.current_quest.val.episode,
                entity.area_id,
            );
        }

        this._selected_entity.val = entity;
    };

    new_quest = (episode: Episode): void => {
        this.set_quest(create_new_quest(episode));
    };

    // TODO: notify user of problems.
    open_file = async (file: File): Promise<void> => {
        try {
            const buffer = await read_file(file);
            const quest = parse_quest(new ArrayBufferCursor(buffer, Endianness.Little));
            this.set_quest(quest && convert_quest_to_model(quest), file.name);
        } catch (e) {
            logger.error("Couldn't read file.", e);
        }
    };

    save_as = (): void => {
        const quest = this.current_quest.val;
        if (!quest) return;

        let default_file_name = this.current_quest_filename.val;

        if (default_file_name) {
            const ext_start = default_file_name.lastIndexOf(".");
            if (ext_start !== -1) default_file_name = default_file_name.slice(0, ext_start);
        }

        let file_name = prompt("File name:", default_file_name);
        if (!file_name) return;

        const buffer = write_quest_qst(convert_quest_from_model(quest), file_name);

        if (!file_name.endsWith(".qst")) {
            file_name += ".qst";
        }

        const a = document.createElement("a");
        a.href = URL.createObjectURL(new Blob([buffer], { type: "application/octet-stream" }));
        a.download = file_name;
        document.body.appendChild(a);
        a.click();
        URL.revokeObjectURL(a.href);
        document.body.removeChild(a);
    };

    id_changed = (event: PropertyChangeEvent<number>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditIdAction(this.current_quest.val, event)).redo();
        }
    };

    name_changed = (event: PropertyChangeEvent<string>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditNameAction(this.current_quest.val, event)).redo();
        }
    };

    short_description_changed = (event: PropertyChangeEvent<string>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditShortDescriptionAction(this.current_quest.val, event)).redo();
        }
    };

    long_description_changed = (event: PropertyChangeEvent<string>): void => {
        if (this.current_quest.val) {
            this.undo.push(new EditLongDescriptionAction(this.current_quest.val, event)).redo();
        }
    };

    translate_entity = (
        entity: QuestEntityModel,
        old_section: SectionModel | undefined,
        new_section: SectionModel | undefined,
        old_position: Vector3,
        new_position: Vector3,
        world: boolean,
    ): void => {
        this.undo
            .push(
                new TranslateEntityAction(
                    entity,
                    old_section,
                    new_section,
                    old_position,
                    new_position,
                    world,
                ),
            )
            .redo();
    };

    rotate_entity = (
        entity: QuestEntityModel,
        old_rotation: Euler,
        new_rotation: Euler,
        world: boolean,
    ): void => {
        this.undo.push(new RotateEntityAction(entity, old_rotation, new_rotation, world)).redo();
    };

    push_create_entity_action = (entity: QuestEntityModel): void => {
        this.undo.push(new CreateEntityAction(entity));
    };

    remove_entity = (entity: QuestEntityModel): void => {
        this.undo.push(new RemoveEntityAction(entity)).redo();
    };

    private async set_quest(quest?: QuestModel, filename?: string): Promise<void> {
        this.undo.reset();

        this._current_area.val = undefined;
        this._selected_entity.val = undefined;

        this._current_quest_filename.val = filename;
        this._current_quest.val = quest;

        if (quest) {
            this._current_area.val = area_store.get_area(quest.episode, 0);

            // Load section data.
            for (const variant of quest.area_variants.val) {
                const sections = await area_store.get_area_sections(quest.episode, variant);
                variant.set_sections(sections);

                for (const object of quest.objects.val.filter(o => o.area_id === variant.area.id)) {
                    try {
                        this.set_section_on_quest_entity(object, sections);
                    } catch (e) {
                        logger.error(e);
                    }
                }

                for (const npc of quest.npcs.val.filter(npc => npc.area_id === variant.area.id)) {
                    try {
                        this.set_section_on_quest_entity(npc, sections);
                    } catch (e) {
                        logger.error(e);
                    }
                }
            }
        } else {
            logger.error("Couldn't parse quest file.");
        }
    }

    private set_section_on_quest_entity = (
        entity: QuestEntityModel,
        sections: SectionModel[],
    ): void => {
        const section = sections.find(s => s.id === entity.section_id.val);

        if (section) {
            entity.set_section(section);
        } else {
            logger.warn(`Section ${entity.section_id.val} not found.`);
        }
    };

    run_current_quest = (): void => {
        const quest = this.current_quest.val;

        if (quest) {
            this.quest_runner.run(quest);
        }
    };

    private log_message_predicate = (msg: LogMessage): boolean => {
        return (
            (this._show_all_log_groups.val || msg.log_group.key === this.log_group.val.key) &&
            msg.log_level.value >= this.log_level.val.value
        );
    };

    public set_log_level(level_arg: LogLevel): void {
        const level = this.log_levels.find(lvl => lvl.value === level_arg.value);

        if (!level) {
            throw new Error(`Invalid argument: No such log level "${level_arg.name}"`);
        }

        this._log_level.val = level;
    }

    public set_log_group(group_arg: LogGroup): void {
        const group = this._log_groups.val.find(grp => grp.key === group_arg.key);

        if (!group) {
            throw new Error(`Invalid argument: No such log group "${group_arg.name}"`);
        }

        this._show_all_log_groups.val = false;
        this._log_group.val = group;
    }

    public show_all_log_groups(): void {
        if (!this._show_all_log_groups.val) {
            this._show_all_log_groups.val = true;
        }
    }

    private add_log_group(name: string): LogGroup {
        const new_group: LogGroup = {
            name: name,
            key: this.log_group_key_counter++,
        };
        this._log_groups.push(new_group);
        return new_group;
    }

    private add_log_message(items: any[], level: LogLevel, group: LogGroup): void {
        this._log_messages.push({
            formatted_timestamp: new Date().toISOString(),
            message_contents: items.join(" "),
            log_level: level,
            log_group: group,
        });
    }

    public get_logger(group_name?: string): Logger {
        let group = this.default_log_group;

        // find existing or create new
        if (group_name !== undefined) {
            const found = this.log_groups.val.find(grp => grp.name === group_name);

            if (found) {
                group = found;
            } else {
                group = this.add_log_group(group_name);
            }
        }

        // logger object bound to log group
        return {
            debug: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Debug, group);
            },
            info: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Info, group);
            },
            warning: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Warning, group);
            },
            error: (...items: any[]): void => {
                this.add_log_message(items, this.log_levels_map.Error, group);
            },
        };
    }
}

export const quest_editor_store = new QuestEditorStore();
