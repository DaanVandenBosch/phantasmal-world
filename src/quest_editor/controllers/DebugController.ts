import { Controller } from "../../core/controllers/Controller";
import { Property } from "../../core/observable/property/Property";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { LogEntry } from "../../core/Logger";
import { LogStore } from "../stores/LogStore";
import { Severity } from "../../core/Severity";
import { property } from "../../core/observable";

export class DebugController extends Controller {
    readonly can_debug: Property<boolean>;
    readonly can_step: Property<boolean>;
    readonly can_stop: Property<boolean>;
    readonly thread_ids: ListProperty<number>;
    readonly debugging_thread_id: Property<number | undefined>;
    readonly active_thread_id: Property<number | undefined>;
    readonly can_select_thread: Property<boolean>;
    readonly log: ListProperty<LogEntry>;
    readonly severity: Property<Severity>;

    constructor(
        gui_store: GuiStore,
        private readonly quest_editor_store: QuestEditorStore,
        private readonly log_store: LogStore,
    ) {
        super();

        this.can_debug = quest_editor_store.current_quest.map(q => q != undefined);

        this.can_step = quest_editor_store.quest_runner.paused;

        this.can_stop = quest_editor_store.quest_runner.running;

        this.thread_ids = quest_editor_store.quest_runner.thread_ids;

        this.debugging_thread_id = quest_editor_store.quest_runner.debugging_thread_id;

        this.active_thread_id = quest_editor_store.quest_runner.active_thread_id;

        this.can_select_thread = quest_editor_store.quest_runner.thread_ids.map(
            ids => ids.length > 0 && quest_editor_store.quest_runner.running.val,
        );

        this.log = log_store.log;
        this.severity = log_store.severity;

        this.disposables(
            gui_store.on_global_keydown(GuiTool.QuestEditor, "F5", this.debug),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F5", this.stop),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F6", this.resume),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F8", this.step_over),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F7", this.step_in),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F8", this.step_out),
        );
    }

    debug = (): void => {
        const quest = this.quest_editor_store.current_quest.val;

        if (quest) {
            this.quest_editor_store.quest_runner.run(quest);
        }
    };

    resume = (): void => {
        this.quest_editor_store.quest_runner.resume();
    };

    step_over = (): void => {
        this.quest_editor_store.quest_runner.step_over();
    };

    step_in = (): void => {
        this.quest_editor_store.quest_runner.step_into();
    };

    step_out = (): void => {
        this.quest_editor_store.quest_runner.step_out();
    };

    stop = (): void => {
        this.quest_editor_store.quest_runner.stop();
    };

    set_severity = (severity: Severity): void => {
        this.log_store.set_severity(severity);
    };

    select_thread = (thread_id: number): void => {
        this.quest_editor_store.quest_runner.set_debugging_thread(thread_id);
    };
}
