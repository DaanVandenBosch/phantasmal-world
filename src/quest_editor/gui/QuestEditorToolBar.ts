import { ToolBar } from "../../core/gui/ToolBar";
import { FileButton } from "../../core/gui/FileButton";
import { Button } from "../../core/gui/Button";
import { quest_editor_store } from "../stores/QuestEditorStore";
import { undo_manager } from "../../core/undo/UndoManager";
import { Select } from "../../core/gui/Select";
import { list_property } from "../../core/observable";
import { AreaModel } from "../model/AreaModel";
import { Icon } from "../../core/gui/dom";
import { DropDown } from "../../core/gui/DropDown";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { area_store } from "../stores/AreaStore";
import { gui_store, GuiTool } from "../../core/stores/GuiStore";

export class QuestEditorToolBar extends ToolBar {
    constructor() {
        const new_quest_button = new DropDown(
            "New quest",
            [Episode.I],
            episode => `Episode ${Episode[episode]}`,
            {
                icon_left: Icon.NewFile,
            },
        );
        const open_file_button = new FileButton("Open file...", {
            icon_left: Icon.File,
            accept: ".qst",
            tooltip: "Open a quest file (Ctrl-O)",
        });
        const save_as_button = new Button("Save as...", {
            icon_left: Icon.Save,
            tooltip: "Save this quest to new file (Ctrl-Shift-S)",
        });
        const undo_button = new Button("Undo", {
            icon_left: Icon.Undo,
            tooltip: undo_manager.first_undo.map(
                action =>
                    (action ? `Undo "${action.description}"` : "Nothing to undo") + " (Ctrl-Z)",
            ),
        });
        const redo_button = new Button("Redo", {
            icon_left: Icon.Redo,
            tooltip: undo_manager.first_redo.map(
                action =>
                    (action ? `Redo "${action.description}"` : "Nothing to redo") +
                    " (Ctrl-Shift-Z)",
            ),
        });
        // TODO: make sure select menu is updated when entity counts change.
        const area_select = new Select<AreaModel>(
            quest_editor_store.current_quest.flat_map(quest => {
                if (quest) {
                    return list_property(
                        undefined,
                        ...area_store.get_areas_for_episode(quest.episode),
                    );
                } else {
                    return list_property<AreaModel>();
                }
            }),
            area => {
                const quest = quest_editor_store.current_quest.val;

                if (quest) {
                    const entity_count = quest.entities_per_area.val.get(area.id);
                    return area.name + (entity_count ? ` (${entity_count})` : "");
                } else {
                    return area.name;
                }
            },
        );
        const run_button = new Button("Run in VM", {
            icon_left: Icon.Play,
            tooltip: "[Experimental] Run the current quest in a virtual machine (F5)",
        });
        const resume_button = new Button("Resume", {
            icon_left: Icon.SquareArrowRight,
            tooltip: "Resume execution"
        });
        const step_over_button = new Button("Step over", {
            icon_left: Icon.LongArrowRight,
            tooltip: "Execute the next line and step over any function calls",
        });
        const step_in_button = new Button("Step in", {
            icon_left: Icon.LevelDown,
            tooltip: "Execute the next line and step inside any function calls"
        });
        const step_out_button = new Button("Step out", {
            icon_left: Icon.LevelUp,
            tooltip: "Execute until outside of current call frame"
        });

        const children = [
            new_quest_button,
            open_file_button,
            save_as_button,
            undo_button,
            redo_button,
            area_select,
        ];

        if (gui_store.feature_active("vm")) {
            children.push(run_button, resume_button, step_over_button, step_in_button, step_out_button);
        }

        super({ children });

        const quest_loaded = quest_editor_store.current_quest.map(q => q != undefined);

        this.disposables(
            new_quest_button.chosen.observe(({ value: episode }) =>
                quest_editor_store.new_quest(episode),
            ),

            open_file_button.files.observe(({ value: files }) => {
                if (files.length) {
                    quest_editor_store.open_file(files[0]);
                }
            }),

            save_as_button.enabled.bind_to(quest_loaded),
            save_as_button.click.observe(quest_editor_store.save_as),

            undo_button.enabled.bind_to(undo_manager.can_undo),
            undo_button.click.observe(() => undo_manager.undo()),

            redo_button.enabled.bind_to(undo_manager.can_redo),
            redo_button.click.observe(() => undo_manager.redo()),

            area_select.enabled.bind_to(quest_loaded),
            area_select.selected.bind_to(quest_editor_store.current_area),
            area_select.selected.observe(({ value: area }) =>
                quest_editor_store.set_current_area(area),
            ),

            run_button.click.observe(quest_editor_store.run_current_quest),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-O", () =>
                open_file_button.click(),
            ),

            gui_store.on_global_keydown(
                GuiTool.QuestEditor,
                "Ctrl-Shift-S",
                quest_editor_store.save_as,
            ),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Z", () => {
                undo_manager.undo();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Shift-Z", () => {
                undo_manager.redo();
            }),

            gui_store.on_global_keydown(
                GuiTool.QuestEditor,
                "F5",
                quest_editor_store.run_current_quest,
            ),
        );

        this.finalize_construction(QuestEditorToolBar.prototype);
    }
}
