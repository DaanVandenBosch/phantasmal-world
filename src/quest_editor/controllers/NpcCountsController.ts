import { Controller } from "../../core/controllers/Controller";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { Property } from "../../core/observable/property/Property";
import { QuestNpcModel } from "../model/QuestNpcModel";
import { npc_data, NpcType } from "../../core/data_formats/parsing/quest/npc_types";
import { property } from "../../core/observable";

export type NameWithCount = { readonly name: string; readonly count: number };

export class NpcCountsController extends Controller {
    readonly npc_counts: Property<readonly NameWithCount[]>;
    readonly unavailable: Property<boolean>;

    constructor(store: QuestEditorStore) {
        super();

        this.unavailable = store.current_quest.map(q => q == undefined);

        this.npc_counts = store.current_quest
            .flat_map(quest => (quest ? quest.npcs : property([])))
            .map(this.update_view);
    }

    private update_view(npcs: readonly QuestNpcModel[]): NameWithCount[] {
        const npc_counts = new Map<NpcType, number>();

        for (const npc of npcs) {
            const val = npc_counts.get(npc.type) || 0;
            npc_counts.set(npc.type, val + 1);
        }

        const extra_canadines = (npc_counts.get(NpcType.Canane) || 0) * 8;

        // Sort by canonical order.
        const sorted_npc_counts = [...npc_counts].sort((a, b) => a[0] - b[0]);

        return sorted_npc_counts.map(([npc_type, count]) => {
            const extra = npc_type === NpcType.Canadine ? extra_canadines : 0;

            return { name: npc_data(npc_type).name, count: count + extra };
        });
    }
}
