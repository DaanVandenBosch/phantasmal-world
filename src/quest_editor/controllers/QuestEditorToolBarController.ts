import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { AreaStore } from "../stores/AreaStore";
import { QuestEditorStore } from "../stores/QuestEditorStore";
import { AreaModel } from "../model/AreaModel";
import { list_property, map } from "../../core/observable";
import { Property } from "../../core/observable/property/Property";
import { undo_manager } from "../../core/undo/UndoManager";
import { Controller } from "../../core/controllers/Controller";
import { Episode } from "../../core/data_formats/parsing/quest/Episode";
import { create_new_quest } from "../stores/quest_creation";
import { read_file } from "../../core/read_file";
import {
    parse_bin_dat_to_quest,
    parse_qst_to_quest,
    Quest,
    write_quest_qst,
} from "../../core/data_formats/parsing/quest";
import { ArrayBufferCursor } from "../../core/data_formats/cursor/ArrayBufferCursor";
import { Endianness } from "../../core/data_formats/Endianness";
import { convert_quest_from_model, convert_quest_to_model } from "../stores/model_conversion";
import { LogManager } from "../../core/Logger";
import { input } from "../../core/gui/dom";
import { basename } from "../../core/util";

const logger = LogManager.get("quest_editor/controllers/QuestEditorToolBarController");

export type AreaAndLabel = { readonly area: AreaModel; readonly label: string };

export class QuestEditorToolBarController extends Controller {
    private quest_filename?: string;

    readonly vm_feature_active: boolean;
    readonly areas: Property<readonly AreaAndLabel[]>;
    readonly current_area: Property<AreaAndLabel>;
    readonly can_save: Property<boolean>;
    readonly can_undo: Property<boolean>;
    readonly can_redo: Property<boolean>;
    readonly can_select_area: Property<boolean>;
    readonly can_debug: Property<boolean>;
    readonly can_step: Property<boolean>;
    readonly can_stop: Property<boolean>;

    constructor(
        gui_store: GuiStore,
        private readonly area_store: AreaStore,
        private readonly quest_editor_store: QuestEditorStore,
    ) {
        super();

        this.vm_feature_active = gui_store.feature_active("vm");

        // Ensure the areas list is updated when entities are added or removed (the count in the
        // label should update).
        this.areas = quest_editor_store.current_quest.flat_map(quest => {
            if (quest) {
                return quest?.entities_per_area.flat_map(entities_per_area => {
                    return list_property<AreaAndLabel>(
                        undefined,
                        ...area_store.get_areas_for_episode(quest.episode).map(area => {
                            const entity_count = entities_per_area.get(area.id);
                            return {
                                area,
                                label: area.name + (entity_count ? ` (${entity_count})` : ""),
                            };
                        }),
                    );
                });
            } else {
                return list_property<AreaAndLabel>();
            }
        });

        this.current_area = map(
            (areas, area) => areas.find(al => al.area == area)!,
            this.areas,
            quest_editor_store.current_area,
        );

        const quest_loaded = quest_editor_store.current_quest.map(q => q != undefined);
        this.can_save = quest_loaded;
        this.can_select_area = quest_loaded;
        this.can_debug = quest_loaded;

        this.can_undo = map(
            (c, r) => c && !r,
            undo_manager.can_undo,
            quest_editor_store.quest_runner.running,
        );

        this.can_redo = map(
            (c, r) => c && !r,
            undo_manager.can_redo,
            quest_editor_store.quest_runner.running,
        );

        this.can_step = quest_editor_store.quest_runner.paused;

        this.can_stop = quest_editor_store.quest_runner.running;

        this.disposables(
            quest_editor_store.current_quest.observe(() => {
                this.quest_filename = undefined;
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-O", () => {
                const input_element = input();
                input_element.type = "file";
                input_element.multiple = true;
                input_element.onchange = () => {
                    if (input_element.files && input_element.files.length) {
                        this.open_files(Array.prototype.slice.apply(input_element.files));
                    }
                };
                input_element.click();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Shift-S", this.save_as),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Z", () => {
                undo_manager.undo();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Ctrl-Shift-Z", () => {
                undo_manager.redo();
            }),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F5", this.debug),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F5", this.stop),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F6", this.resume),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F8", this.step_over),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "F7", this.step_in),

            gui_store.on_global_keydown(GuiTool.QuestEditor, "Shift-F8", this.step_out),
        );
    }

    create_new_quest = async (episode: Episode): Promise<void> =>
        this.quest_editor_store.set_current_quest(create_new_quest(this.area_store, episode));

    // TODO: notify user of problems.
    open_files = async (files: File[]): Promise<void> => {
        try {
            let quest: Quest | undefined;

            const qst = files.find(f => f.name.toLowerCase().endsWith(".qst"));

            if (qst) {
                const buffer = await read_file(qst);
                quest = parse_qst_to_quest(new ArrayBufferCursor(buffer, Endianness.Little));
                this.quest_filename = qst.name;
            } else {
                const bin = files.find(f => f.name.toLowerCase().endsWith(".bin"));
                const dat = files.find(f => f.name.toLowerCase().endsWith(".dat"));

                if (bin && dat) {
                    const bin_buffer = await read_file(bin);
                    const dat_buffer = await read_file(dat);
                    quest = parse_bin_dat_to_quest(
                        new ArrayBufferCursor(bin_buffer, Endianness.Little),
                        new ArrayBufferCursor(dat_buffer, Endianness.Little),
                    );
                    this.quest_filename = bin.name || dat.name;
                }
            }

            if (!quest) {
                logger.error("Couldn't parse quest file.");
            }

            await this.quest_editor_store.set_current_quest(
                quest && convert_quest_to_model(this.area_store, quest),
            );
        } catch (e) {
            logger.error("Couldn't read file.", e);
        }
    };

    set_area = ({ area }: AreaAndLabel): void => {
        this.quest_editor_store.set_current_area(area);
    };

    save_as = (): void => {
        const quest = this.quest_editor_store.current_quest.val;
        if (!quest) return;

        const default_file_name = this.quest_filename && basename(this.quest_filename);

        let file_name = prompt("File name:", default_file_name);
        if (!file_name) return;

        const buffer = write_quest_qst(convert_quest_from_model(quest), file_name);

        if (!file_name.endsWith(".qst")) {
            file_name += ".qst";
        }

        const a = document.createElement("a");
        a.href = URL.createObjectURL(new Blob([buffer], { type: "application/octet-stream" }));
        a.download = file_name;
        document.body.appendChild(a);
        a.click();
        URL.revokeObjectURL(a.href);
        document.body.removeChild(a);
    };

    debug = (): void => {
        const quest = this.quest_editor_store.current_quest.val;

        if (quest) {
            this.quest_editor_store.quest_runner.run(quest);
        }
    };

    resume = (): void => {
        this.quest_editor_store.quest_runner.resume();
    };

    step_over = (): void => {
        this.quest_editor_store.quest_runner.step_over();
    };

    step_in = (): void => {
        this.quest_editor_store.quest_runner.step_into();
    };

    step_out = (): void => {
        this.quest_editor_store.quest_runner.step_out();
    };

    stop = (): void => {
        this.quest_editor_store.quest_runner.stop();
    };
}
