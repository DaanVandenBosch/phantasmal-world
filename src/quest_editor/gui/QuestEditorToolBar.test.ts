/**
 * @jest-environment jsdom
 */
import { QuestEditorToolBarController } from "../controllers/QuestEditorToolBarController";
import { QuestEditorToolBar } from "./QuestEditorToolBar";
import { GuiStore } from "../../core/stores/GuiStore";
import { create_area_store } from "../../../test/src/quest_editor/stores/store_creation";
import { QuestEditorStore } from "../stores/QuestEditorStore";

test("Renders correctly.", () => {
    const gui_store = new GuiStore();
    const area_store = create_area_store();
    const quest_editor_store = new QuestEditorStore(gui_store, area_store);
    const tool_bar = new QuestEditorToolBar(
        new QuestEditorToolBarController(gui_store, area_store, quest_editor_store),
    );

    expect(tool_bar.element).toMatchSnapshot();
});
