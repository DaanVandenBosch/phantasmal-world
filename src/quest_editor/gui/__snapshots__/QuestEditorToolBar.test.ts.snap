// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly. 1`] = `
<div
  class="core_ToolBar"
  style="height: 33px;"
>
  <div
    class="core_DropDown"
  >
    <button
      class="core_Button"
    >
      <span
        class="core_Button_inner"
      >
        <span
          class="core_Button_left"
        >
          <span
            class="fas fa-file-medical"
          />
        </span>
        <span
          class="core_Button_center"
        >
          New quest
        </span>
        <span
          class="core_Button_right"
        >
          <span
            class="fas fa-caret-down"
          />
        </span>
      </span>
    </button>
    <div
      class="core_Menu"
      hidden=""
      tabindex="-1"
    >
      <div
        class="core_Menu_inner"
      >
        <div
          data-index="0"
        >
          Episode I
        </div>
      </div>
    </div>
  </div>
  <label
    class="core_FileButton core_Button"
    title="Open a quest file (Ctrl-O)"
  >
    <span
      class="core_FileButton_inner core_Button_inner"
    >
      <span
        class="core_FileButton_left core_Button_left"
      >
        <span
          class="fas fa-file"
        />
      </span>
      <span
        class="core_Button_center"
      >
        Open file...
      </span>
    </span>
    <input
      accept=".qst"
      class="core_FileButton_input core_Button_inner"
      type="file"
    />
  </label>
  <button
    class="core_Button disabled"
    disabled=""
    title="Save this quest to new file (Ctrl-Shift-S)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span
          class="fas fa-save"
        />
      </span>
      <span
        class="core_Button_center"
      >
        Save as...
      </span>
    </span>
  </button>
  <button
    class="core_Button disabled"
    disabled=""
    title="Nothing to undo (Ctrl-Z)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span
          class="fas fa-undo"
        />
      </span>
      <span
        class="core_Button_center"
      >
        Undo
      </span>
    </span>
  </button>
  <button
    class="core_Button disabled"
    disabled=""
    title="Nothing to redo (Ctrl-Shift-Z)"
  >
    <span
      class="core_Button_inner"
    >
      <span
        class="core_Button_left"
      >
        <span
          class="fas fa-redo"
        />
      </span>
      <span
        class="core_Button_center"
      >
        Redo
      </span>
    </span>
  </button>
  <div
    class="core_Select disabled"
  >
    <button
      class="core_Button disabled"
      disabled=""
    >
      <span
        class="core_Button_inner"
      >
        <span
          class="core_Button_center"
        >
           
        </span>
        <span
          class="core_Button_right"
        >
          <span
            class="fas fa-caret-down"
          />
        </span>
      </span>
    </button>
    <div
      class="core_Menu"
      hidden=""
      tabindex="-1"
    >
      <div
        class="core_Menu_inner"
      />
    </div>
  </div>
</div>
`;
